client/src/controllers/LoaderController.test.js

```diff
@@ -0,0 +1,99 @@
+import { Application } from '@hotwired/stimulus';
+import { LoaderController } from './LoaderController';
+
+describe('LoaderController', () => {
+  beforeEach(() => {
+    document.body.innerHTML = `
+            <button
+                type="submit"
+                class="button button-longrunning"
+                data-controller="w-loader"
+                data-w-loader-active-class="button-longrunning-active"
+                data-w-loader-active-value="Signing in…"
+                data-w-loader-duration-seconds-value="40"
+                data-action="w-loader#activate"
+            >
+                <svg>...</svg>
+                <em data-w-loader-target="label" id="em-el">Sign in</em>
+            </button>
+        `;
+    Application.start().register('w-loader', LoaderController);
+  });
+
+  afterEach(() => {
+    document.body.innerHTML = '';
+    jest.clearAllMocks();
+    jest.clearAllTimers();
+  });
+
+  it('should not execute the rest of the code if the form is not valid', () => {
+    const form = document.createElement('form');
+    const button = document.querySelector('.button-longrunning');
+    let returnCalled = false;
+
+    form.noValidate = false;
+    form.checkValidity = jest.fn().mockReturnValue(false);
+    button.closest = jest.fn().mockReturnValue(form);
+
+    if (
+      form &&
+      form.checkValidity &&
+      !form.noValidate &&
+      !form.checkValidity()
+    ) {
+      returnCalled = true;
+      return;
+    }
+
+    expect(returnCalled).toBe(true);
```

Comment:
Try to do your assertions against the HTML and the button click. 

This test does not actually click the button and does not appear to be testing what you want. 

E.g. 

```js
const onClick = jest.fn();
button.addEventListener('click', onClick);
button.dispatchEvent(new CustomEvent('click'));
await Promise.resolve(): // check syntax but we just want to do a rendering pass
expect(button.disabled).toEqual(false);
expect(onClick).toHaveBeenCalledTimes(1);
```

You are checking the button click still works AND the button does not get disabled.